doc/sphinx/proof-engine/ltac.rst

```diff
@@ -28,14 +29,27 @@ especially about its foundations, please refer to :cite:`Del00`.
 Syntax
 ------
 
-The syntax of the tactic language is given below. See Chapter
-:ref:`gallinaspecificationlanguage` for a description of the BNF metasyntax used
-in these grammar rules. Various already defined entries will be used in this
-chapter: entries :token:`num`, :token:`int`, :token:`ident`
-:token:`qualid`, :token:`term`, :token:`cpattern` and :token:`tactic`
-represent respectively natural and integer numbers,
-identifiers, qualified names, Coq terms, patterns and the atomic
-tactics described in Chapter :ref:`tactics`.
+The syntax of the tactic language is given below.
+
+The main entry of the grammar is :n:`@ltac_expr`, which is used in proof mode
+as well as to define new tactics with the :cmd:`Ltac` command.
+
+The grammar uses multiple :n:`ltac_expr*` nonterminals to express how subexpressions
+are grouped when they're not fully parenthesized.  For example, in many programming
+languages, `a*b+c` is interpreted as `(a*b)+c` because `*` has
+higher precedence than `+`.  Usually `a/b/c` is given the :gdef:`left associative`
+interpretation `(a/b)/c` rather than the :gdef:`right associative` interpretation
+`a/(b/c)`.
+
+In Coq, the expression :n:`try repeat @tactic__1 || @tactic__2; @tactic__3; @tactic__4`
+is interpreted as :n:`((try (repeat (@tactic__1 || @tactic__2)); @tactic__3)); @tactic__4`
+because `||` is part of :token:`ltac_expr2`, which has higher precedence than
+:tacn:`try` and :tacn:`repeat` (at the level of :token:`ltac_expr3`), which
+in turn have higher precedence than `;`, which is left-associative and part
+of :token:`ltac_expr4`.  (A *lower* number in the nonterminal name means *higher* precedence
+in this grammar.)
+
+The constructs in :token:`ltac_expr` are :term:`left associative`.
 
 The syntax of :production:`cpattern` is
```

Comment:
`cpattern` was incorrect even before I began updating the chapter.  `context` uses `Constr.lconstr`, which I've collapsed into `term`.  But it seems useful to keep this distinction in the doc.  They only need a sentence each to explain and it will be clearer.  I also want to bring back `uconstr`.  Maybe call these `term_pattern` and `untyped_term`?

FYI there is a `cpattern` in the SSR grammar, seems clear that that's something else.

Reply:
Also, later in this paragraph, I don't see how the grammar accepts `%@?ident` in "an applicative pattern of the form %@?ident ident1 … identn".  What should this be?

Reply:
Another place where we could have done this kind of distinction but didn't do it is the syntax of `Search`...

> Also, later in this paragraph, I don't see how the grammar accepts `%@?ident` in "an applicative pattern of the form %@?ident ident1 … identn". What should this be?

The `%` is spurious. Someone probably thought that `@` needed some escaping.

Reply:
Removed the %.  You asked to omit this production from term10 a while ago.  When and where do we re-add it?

`| "@" "?" ident LIST1 ident`

Reply:
Good point. We should definitely spend time explaining this pattern syntax in detail. First of all, we should clarify what is Ltac-specific and what is more generic. Everything that is generic could be documented in the same section as existential variables IMHO. To be left as later work, otherwise we'll never manage to merge this PR in a reasonable time frame.

Reply:
OK, I'll leave a todo in the file linking back to here.

Also, BTW, what do you think about replacing `...` (3 periods) with `…` (ellipsis character).  I think that might be a bit clearer in, say, the tactic index?

Reply:
For `cpattern`, I think the way to fix this (in this PR) is to rename `lconstr` as `term_pattern` and replace `cpattern` in the text with `term_pattern`.  And maybe also move the note somewhere else.

OTOH, when I try that change I see `term_pattern` in unexpected places--perhaps minor bugs in the grammar?  For example:

```
.. cmd:: {| Definition | Example } @ident_decl @def_body
   :name: Definition; Example

   .. insertprodn def_body def_body

   .. prodn::
      def_body ::= {* @binder } {? : @type } := {? @reduce } @term_pattern
      | {* @binder } : @type
```

WDYT?

Reply:
> Also, BTW, what do you think about replacing `...` (3 periods) with `…` (ellipsis character). I think that might be a bit clearer in, say, the tactic index?

Sure.

> OTOH, when I try that change I see `term_pattern` in unexpected places--perhaps minor bugs in the grammar? For example:

I thought `lconstr` and `operconstr200` where the same thing (i.e. what we call `term`). If that's so, then it is not surprising that you see this. More generally, the issue AFAICT is simply that the grammar of terms and patterns is shared. So there is no easy way around that, and I would simply remove `cpattern` and explain that `term` is to be understood as a pattern in the constructs that used it. Cf. my initial suggesstion.

Reply:
> I thought lconstr and operconstr200 where the same thing (i.e. what we call term).

That's correct.

I expected that terms that contain `_` and `? ident` won't type check--is that right?

> So there is no easy way around that, and I would simply remove cpattern and explain that term is to be understood as a pattern in the constructs that used it.

It would be foolish to describe it syntactically, but it's simple to add a production `cpattern ::= term` or `term_pattern ::= term`, use the new nonterminal where appropriate and then describe the semantic limitation in text.  Sort of like what we did for defining `type`.

Probably the natural place to put this is with the definition of `term_evar`, though I could put it elsewhere.  WDYT?

Reply:
> I expected that terms that contain `_` and `? ident` won't type check--is that right?

That is not correct. If Coq can guess the value of the hole, it will type check. Cf. https://coq.github.io/doc/master/refman/language/extensions/implicit-arguments.html#casual-use-of-implicit-arguments

> It would be foolish to describe it syntactically, but it's simple to add a production `cpattern ::= term` or `term_pattern ::= term`, use the new nonterminal where appropriate and then describe the semantic limitation in text. Sort of like what we did for defining `type`.

That would feel OK to me in principle, but we are also facing the issue that sometimes such patterns are `lconstr` / `term` and sometimes they are `constr` / `one_term`.

> Probably the natural place to put this is with the definition of `term_evar`, though I could put it elsewhere. WDYT?

That was also my idea, but I'd like to clarify first whether the pattern language is exactly the same across Coq or is slightly different in the case of Ltac constructs. Therefore, for this PR, I would put the definition close to where it is used.

Reply:
OK, I moved it match term.  We can revisit in the future.